### PR TITLE
test: Replace hard-coded hex tx with class in test framework

### PR DESCRIPTION
**Pull request by stevenroose**
Came across these breaking Elements.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
This is used in some "hot loops", so making a benchmark to prove it is not significantly slowing down the function call would help.

**Comment by stevenroose** [comment]
I don't know if this helps:
```
$ cat exec.sh 
#!/bin/sh
./test/functional/mempool_limit.py && ./test/functional/mining_prioritisetransaction.py 
$ time ./exec.sh 
2019-01-29T22:22:07.529000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_vn7gv6oi
2019-01-29T22:22:07.814000Z TestFramework (INFO): Check that mempoolminfee is minrelytxfee
2019-01-29T22:22:08.241000Z TestFramework (INFO): Create a mempool tx that will be evicted
2019-01-29T22:22:09.816000Z TestFramework (INFO): The tx should be evicted by now
2019-01-29T22:22:09.829000Z TestFramework (INFO): Check that mempoolminfee is larger than minrelytxfee
2019-01-29T22:22:09.834000Z TestFramework (INFO): Create a mempool tx that will not pass mempoolminfee
2019-01-29T22:22:09.906000Z TestFramework (INFO): Stopping nodes
2019-01-29T22:22:10.110000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_vn7gv6oi on exit
2019-01-29T22:22:10.110000Z TestFramework (INFO): Tests successful
2019-01-29T22:22:10.215000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_5zgic6c6
2019-01-29T22:22:12.664000Z TestFramework (INFO): Assert that prioritised transaction was mined
2019-01-29T22:22:12.797000Z TestFramework (INFO): Assert that de-prioritised transaction is still in mempool
2019-01-29T22:22:12.905000Z TestFramework (INFO): Assert that prioritised free transaction is accepted to mempool
2019-01-29T22:22:13.070000Z TestFramework (INFO): Stopping nodes
2019-01-29T22:22:13.372000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_5zgic6c6 on exit
2019-01-29T22:22:13.372000Z TestFramework (INFO): Tests successful
./exec.sh  3.88s user 0.49s system 72% cpu 5.997 total
$ gco master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
$ time ./exec.sh
2019-01-29T22:22:22.500000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_c_ph_h1g
2019-01-29T22:22:22.772000Z TestFramework (INFO): Check that mempoolminfee is minrelytxfee
2019-01-29T22:22:23.222000Z TestFramework (INFO): Create a mempool tx that will be evicted
2019-01-29T22:22:24.519000Z TestFramework (INFO): The tx should be evicted by now
2019-01-29T22:22:24.521000Z TestFramework (INFO): Check that mempoolminfee is larger than minrelytxfee
2019-01-29T22:22:24.522000Z TestFramework (INFO): Create a mempool tx that will not pass mempoolminfee
2019-01-29T22:22:24.578000Z TestFramework (INFO): Stopping nodes
2019-01-29T22:22:24.831000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_c_ph_h1g on exit
2019-01-29T22:22:24.831000Z TestFramework (INFO): Tests successful
2019-01-29T22:22:24.947000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_nf8kqb1j
2019-01-29T22:22:27.376000Z TestFramework (INFO): Assert that prioritised transaction was mined
2019-01-29T22:22:27.520000Z TestFramework (INFO): Assert that de-prioritised transaction is still in mempool
2019-01-29T22:22:27.600000Z TestFramework (INFO): Assert that prioritised free transaction is accepted to mempool
2019-01-29T22:22:27.752000Z TestFramework (INFO): Stopping nodes
2019-01-29T22:22:27.905000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_nf8kqb1j on exit
2019-01-29T22:22:27.905000Z TestFramework (INFO): Tests successful
./exec.sh  3.70s user 0.51s system 75% cpu 5.546 total
```

**Comment by instagibbs** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/15282/commits/03f7a8f392c0f5574ca2dbc19e7a16d32dc97312

Extremely minor slowdown imo for great increase in test clarity.

File: test/functional/test_framework/util.py [hunk 1]
```diff
@@ -516,32 +517,34 @@ def gen_return_txouts():
...
+    txouts = []
+    from .messages import CTxOut
     for k in range(128):
-        # add txout value
-        txouts = txouts + "0000000000000000"
-        # add length of script_pubkey
-        txouts = txouts + "fd0402"
-        # add script_pubkey
-        txouts = txouts + script_pubkey
+        txout = CTxOut()
```
Comment (instagibbs) [inline test/functional/test_framework/util.py]: You could take out everything from this hot loop except the `txtouts.append` line to see if we gain back that 0.18 secs?

Comment (stevenroose) [reply test/functional/test_framework/util.py]: Strangely enough this made it 0.2 seconds slower on my machine :/ So I think it's better as-is.

**Comment by stevenroose** [comment]
It seems like https://github.com/bitcoin/bitcoin/pull/14691 is making this change obsolete.

**Comment by laanwj** [comment]
@MarcoFalke given the benchmarks, are you ok with this ?

**Comment by MarcoFalke** [comment]
Sure but there is also [#14691](https://drahtbot.github.io/bitcoin_core_issue_redirect/r/14691.html) (tests: Speedup feature_pruning test and refactor big transaction logic by @conscott) which look promising

**Comment by MarcoFalke** [comment]
@stevenroose Is this still relevant?

**Comment by fanquake** [comment]
Ping @stevenroose.

**Comment by stevenroose** [comment]
Let me try rebase this.

**Comment by stevenroose** [comment]
@fanquake @MarcoFalke rebased. Up to you guys whether it's relevant, of course. I think it's an improvement.

**Comment by MarcoFalke** [comment]
ACK 8f250ab7882a852f1b1947cef4837d2de5ca6913

**Comment by practicalswift** [comment]
Concept ACK

File: test/functional/test_framework/util.py [hunk 2]
```diff
@@ -515,31 +516,33 @@ def gen_return_txouts():
...
         inputs = [{"txid": t["txid"], "vout": t["vout"]}]
         outputs = {}
         change = t['amount'] - fee
         outputs[addr] = satoshi_round(change)
         rawtx = node.createrawtransaction(inputs, outputs)
-        newtx = rawtx[0:92]
-        newtx = newtx + txouts
-        newtx = newtx + rawtx[94:]
+        tx = CTransaction()
+        tx.deserialize(BytesIO(hex_str_to_bytes(rawtx)))
```
Comment (instagibbs) [inline test/functional/test_framework/util.py]: please don't let this block merging: I think you can use `FromHex(CTransaction(), rawtx)` here for a short one-liner

**Comment by instagibbs** [comment]
utACK https://github.com/bitcoin/bitcoin/commit/8f250ab7882a852f1b1947cef4837d2de5ca6913

**Pull request merged**